engines/dradis-api/lib/dradis/ce/api/warden_strategy.rb

```diff
@@ -32,11 +38,26 @@ def authenticate!
     end
 
     private
+    def token
+      @token ||= ActionController::HttpAuthentication::Token.token_and_options(ActionDispatch::Request.new(env)).first rescue nil
+    end
 
     def auth
       @auth ||= Rack::Auth::Basic::Request.new(env)
     end
 
+    # Taken from [Devise](https://github.com/plataformatec/devise).
+    # constant-time comparison algorithm to prevent timing attacks
+    def secure_compare(a)
+      b = token
+      return false if a.blank? || b.blank? || a.bytesize != b.bytesize
+      l = a.unpack "C#{a.bytesize}"
+
+      res = 0
+      b.each_byte { |byte| res |= byte ^ l.shift  }
+      res == 0
+    end
+
```

Comment:
We weren't using this before with token authentication in Pro, why add it now?

Reply:
This was ported from the Pro warden strategy and used there too, so this isn't anything new.